Extend multi-window support to Electron (#11642)

tsmaeder opened the pull request:
#### What it does
Fixes #11642

The main change is to prevent the secondary window from closing until the extracted widget is removed from the window. This includes waiting until any close handling (including dialogs) are finished.

Contributed on behalf of ST Microelectronics.

#### How to test
Since this just enables the functionality from https://github.com/eclipse-theia/theia/pull/11048 for electron, the "how to test" from that PR applies.

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

lucas-koehler reviewed (changes requested):
Hi @tsmaeder ,

I tested the changes and, unfortunately,  still noticed a few issues.

First, the updated `examples/electron/tsconfig.json` should be committed with the reference to the `secondary-window` extension.

Second, I could still get `IllegalAccess` errors sometimes. Albeit, I feel their frequency has gone down. To test this I used the cat coding extension as a simple widget and the [draw.io integration](https://marketplace.visualstudio.com/items?itemName=hediet.vscode-drawio) as a custom editor.

On Linux (Ubuntu 20.04), widgets could be reopened even if an `IllegalAccess` was reported. It seemed that this was more likely with activated autoSave on the custom editor.

On Windows 11 (VM), widgets could sometimes not be re-opened after an `IllegalAccess` was reported.

On both systems, closing the draw io custom editor via the `Open Editors` view did close the widget in the external window but the window stayed open as an empty shell.

Besides these issues, I also discovered positive news about the auto save:
With the draw io extension, auto save was applied in the editor. Meaning, when the setting was activated, the editor automatically persisted changes. Furthermore, the save dialog worked when closing a dirty draw io editor in an external window: The closing of the widget and window were stopped until the dialog was closed. The dialog was opened in the main window. However, I think this is another issue: https://github.com/eclipse-theia/theia/issues/11647

tsmaeder commented:
Hi @lucas-koehler I don't ever get any "illegal access" with the code in this PR on Window 10. Do you have a scenario and frequencies if not repeatable?

lucas-koehler commented:
Hi @tsmaeder,
I have two sencarios two reproduce this:

### Cat Coding

1. Repeatedly opening and closing the cat coding widget in a secondary window. Frequency was after about 20 tries. I repeated this 2 times.

### Custom editor

1. Enable autosave
2. opening a draw io diagram, change something, immediately close window
3. this seems to work quite reliably. However, the `Illegal Access` seems to have different consequences as the widget can be re-opened.

https://user-images.githubusercontent.com/6959840/191026528-801c3a04-f486-4bbe-864d-897cee367384.mp4

tsmaeder commented:
Thx @lucas-koehler. Am I correct that both scenarios are with "autosave" on?

tsmaeder commented:
I'm having trouble reproducing the problem reliably: it just seems to randomly happen. When I start adding console output to track the problem, the problem goes away. That means that I can't just try out stuff and see if if fixes the problem, since I can't verify a fix if I can't reasonably provoke the problem. <sigh>

paul-marechal commented:
When testing on Windows I sometime get in a state where the secondary windows opened is empty and I get `Uncaught illegal access` when I close the Window. I also noticed that the Theia/PhosphorJS widget seems to be lingering from this point onward, preventing the webview from opening again (I tested using the `Hex Editor` extension).

I cannot find a surefire way to reproduce it but I'm also able to end up in a state where the webview widget is leaked which prevents me from opening the webview, and this without triggering any error. It seems to somehow remove parts of the editor title bar which is odd:

![image](https://user-images.githubusercontent.com/14182238/191380107-97c0bd50-8b4f-432c-b1d4-71449c5154f2.png)

By "widget leaked" I mean that the widget is found in the following piece of logic despite not being on visible in Theia:

[custom-editor-opener.tsx#L68](https://github.com/tsmaeder/theia/blob/1b824ebdcac82d3e165597e4387cf9f9c7e6520d/packages/plugin-ext/src/main/browser/custom-editors/custom-editor-opener.tsx#L68)

Only when `widget` is `undefined` does it seem to show the webview.

tsmaeder commented:
Unfortunately, it's quite hard to reproduce this issue for me: sometimes I can try for 15 minutes and not reproduce it, sometimes it happens right away. After lengthy experimentation, I've found the following scenario that reproduces the issue quite reliably. This scenario uses the drawio custom editor

1. create two files x.drawio and bar.drawio
2. Open x.drawio, then bar.drawio
3. Move x.drawio to a secondary window
4. Use the "Open Editors" view in the navigator to bring the secondary window to the front
5. Move a figure in the x.drawio editor and immediately close the secondary window
6. Observe the "illegal access" in the main window browser log

tsmaeder commented:
Here's a recording of the scenario:

https://user-images.githubusercontent.com/13163770/191441192-8389eedf-92d3-4b60-bc76-d66514bff57a.mp4

tsmaeder commented:
I've added code to remove the 'close' handler from the electron window when it is hit. For me, that seems to fix the problem, but @lucas-koehler can still reproduce the issue. That the electron-remote stuff is involved somehow would make sense, though. A thing to try would be to remove the electron-remote stuff and see if that fixes the problem.

tsmaeder commented:
I'll go back to working on https://github.com/eclipse-theia/theia/issues/11643 for the time being.

tsmaeder commented:
I had some further insights on this problem:

* Clean handling of saving dirty editors will not be possible (using the save/don't save/cancel dialog) will not be possible in the browser case: we cannot really prevent browser windows from closing: the best we can do is pop up the native dialog asking whether to "leave the page without saving", same as we do when we close the main window. This prevents the approach of doing a proper remove of the widget and postponing the close after close actions have finished.



* The approach should work in Electron, though, by adding an `on('close')` handler.
  * Update: registering a 'close' handler with electronRemote does not seem to prevent closing of the browser window, even if we do a "preventDefault". This needs debugging.

tsmaeder commented:
Confirmed that the close handler works fine when attaching it to the newly create window in the electron main process. So either we're doing something wrong in the current logic or we need to figure out how to call the proper pre-close logic from the electron main process.

tsmaeder committed:
Add secondary window support for Electron.

Fixes #11642

The main change is to prevent the secondary window from closing until
the extracted widget is removed from the window. This includes waiting
until any close handling (including dialogs) are finished.
Since we cannot reliably prevent closing windows in the browser case,
we either save or discard unsaved changes according to the autosave
settings

Contributed on behalf of ST Microelectronics.

Signed-off-by: Thomas Mäder <t.s.maeder@gmail.com>

tsmaeder commented:
@lucas-koehler I've updated with my latest changes and rebased on origin/master. Could you have another look?

lucas-koehler reviewed (commented):
Browser works as expected given the limitations.
Unfortunately, the IllegalAccess is back using the proven reproduction approach with drawio and two editors.
Idk why that happens as it was not reproducible for me back when I tested the https://github.com/tsmaeder/theia/commits/11642_electron_secondary_window_2 branch a week or so ago.

It was reproducible after one or two tries.
Additional info:
Ubuntu 20.04
Default plugins + drawio editor and cat coding
Node 16.15.1, yarn 1.22.15
I even cloned a new repo with only your branch to avoid some left over pollution of the repo.

lucas-koehler commented on packages/core/src/browser/secondary-window-handler.ts:
```diff
@@ -139,13 +139,32 @@ export class SecondaryWindowHandler {
+                    };
+                }
+
+                await this.applicationShell.closeWidget(widget.id, closeOptions);
+                widget.isClosing = false;
+                return widget.isDisposed;
```
`await this.applicationShell.closeWidget(widget.id, closeOptions);` closes the secondary window that the widget resides in with the changes at application-shell.ts https://github.com/eclipsesource/theia/pull/49/files#diff-cb4b87cc3dc274963d24f6fd5682ad9c89cd51d94726100eb2e3c92421e5c171R1565-R1566

Afterwards, the widget is accessed to set its `isClosing` flag. Maybe this could cause problems with `IllegalAccess`?
This is especially intereseting because this was not the case on the last known working state at https://github.com/cdamus/theia/commit/2b3d733f8029b8e74a8e471192c6abee8f36e79f

The changes stem from the browser save handling as far as I can see but affect the electron implementation in this way, too. It is called here: https://github.com/eclipsesource/theia/pull/49/files#diff-563e180229ce37170c43089f2a51e72ab90eead76b88b8a4300365d8373f93c3R51

tsmaeder replied:
All I can say is that this _shouldn't_ be a problem, since the widget is an object that lives in the context of the main browser window. But it's a diff I should look at.

tsmaeder commented:
I went through the compiled PhosphorJS code (`widget.js`) and noticed that `widget.id` is actually a computed property that accesses the node. Hmh...that is something to look at.

tsmaeder commented:
Another note: I can reproduce the "illegal access" as well with terminal windows (though not as reliably), so this does _not_ seem to have anything to do with the particularities of webviews.

tsmaeder commented:
For what it's worth, I got a case of "illegal access" from a section of the electron renderer windows init code, from the section that assigned a window error handler inside the "nodeIntegration=true" branch of execution.

tsmaeder commented:
So here's what I tried:

1. Rebased the original tested change to commit before terminal secondary windows: => could not reproduce
2. Rebased to commit introducing terminal secondary windows => could reproduce.

As always, this could be conincidence: I'll cross-check.

tsmaeder commented:
I went back to the state rebased on before the external terminals and I can (after around 20 tries) reproduce the issue. Back to square one.

tsmaeder commented:
Bad news: I retested the original version that we thought was error-free and was able to provoke the "illegal access" after around 20 tries. The fact seems to be that if we remove bogus accesses (like the one where we keep references to the closed window's document when we update the theme), we can make the errors less likely. This kind of gives me hope that the error is something we cause. On the other hand, we're not really any closer to finding fixing this problem for good.

tsmaeder commented:
Another candidate for causing the problem is the "open editors" widget: refreshing a tree (by assigning to the root) is always async with no ability to wait for the refresh to end.

tsmaeder commented:
Another thing to try would be to make the the creation of the secondary window loop around to the main electron thread cleanly instead of using electron-remote.

tsmaeder committed:
Some changes to prevent "Illegal access"

Signed-off-by: Thomas Mäder <t.s.maeder@gmail.com>

eneufeld commented on packages/core/src/electron-main/theia-electron-window.ts:
```diff
@@ -75,6 +82,37 @@ export class TheiaElectronWindow {
+                            // to prevent "illegal access" errors. Unfortunately, it is unclear what 
+                            // exactly causes the errors.
+                            this._window.focus();
+                            setTimeout(() => {
```
I suggest to replace lines 101 to 104 with a simple
`newWindow.destroy();`
When reacting to the newWindow `close` event, the default propagation of event is stopped in order to check that the second window is safe to be closed. After the check the window was calling `close` which again triggers the close event, what is not needed. So `destroy` is the more accurate method to use.
With the change most/all? `Illegal access` errors are gone.

eneufeld replied:
That was to early, unfortunately the error can still be reproduced after a while.

eneufeld commented on packages/core/src/electron-main/theia-electron-window.ts:
```diff
@@ -75,6 +82,37 @@ export class TheiaElectronWindow {
+                            // to prevent "illegal access" errors. Unfortunately, it is unclear what 
+                            // exactly causes the errors.
+                            this._window.focus();
+                            setTimeout(() => {
+                                closingState = ClosingState.readyToClose;
+                                newWindow.close();
+                            }, 100);
```
```suggestion
                            newWindow.destroy();
```

eneufeld commented:
I think there is a memory leak when detaching the windows. 
Here is a heaptimeline https://drive.google.com/file/d/1Xp3L0vxjzWvJHY589s-UKBzu7q_4ueAH/view?usp=share_link . Unfortunately I can't directly attach it as the file is to big.

tsmaeder commented:
I am not confident that I'll get anywhere without being able to look at local variables. I'm not fit enough with my native debugging to look at the stack in hex, so it's important to get a build at `symbol_level = 2`. I'll have to try on Linux, I guess.
